build: adding a script to compare commits in master and stable branches

AndrewKushnir opened the pull request:
Adding a script that compares commits in master and patch branches and finds a delta between them. This is useful for release reviews, to make sure all the necessary commits are included into the patch branch and there is no discrepancy.

This script was inspired by the [compare-master-to-stable.js](https://github.com/angular/angular.js/blob/master/scripts/compare-master-to-stable.js) one that was used in AngularJS.


## PR Type
What kind of change does this PR introduce?

- [x] Other... Please describe: a tool to compare commits in master and stable branches


## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

AndrewKushnir committed:
build: adding a script to compare commits in master and stable branches

Adding a script that compares commits in master and patch branches and finds a delta between them. This is useful for release reviews, to make sure all the necessary commits are included into the patch branch and there is no discrepancy.

kara commented on scripts/compare-master-to-stable.js:
```diff
@@ -0,0 +1,122 @@
+console.log(`
+Comparing commits in ranges: (${from}...master) and (${from}...${to}).
+
+***** Only in MASTER *****
```
When I run this script, I see commits in this section that are also on patch, e.g.  afdd405995. Is this intentional?

AndrewKushnir replied:
I've updated the script based on our discussion with @josephperrott, it should be more accurate now.

AndrewKushnir committed:
fixup! build: adding a script to compare commits in master and stable branches

kara commented on scripts/compare-master-to-stable.js:
```diff
@@ -0,0 +1,121 @@
+console.log(`
+Comparing branches "${branch}" and master.
+
+***** Only in MASTER *****
```
I'm still seeing some commits under "ONLY IN MASTER" that are also on 9.0.x branch. e.g.

ci: reenable draft mode conditioning for pullapprove (#35396)
ci: add verification of the pullapprove config (#35060) …

kara reviewed (changes requested):
Can you run the script and sanity check the build/ci commits that are in ONLY IN MASTER? It doesn't look right to me.

AndrewKushnir commented:
Hi @kara, I updated `master` and `9.0.x` branches locally, ran a script after that and I don't see these two commits that you refer to in a diff. May be try fetching the latest `9.0.x` and run the script once again to see if the diff is correct now? Thank you.

kara commented:
@AndrewKushnir Updating my 9.0.x branch fixed the issue. But it concerns me that it's so easy to get into this state by mistake. Could we throw if we notice the branch is out of date?

AndrewKushnir committed:
fixup! build: adding a script to compare commits in master and stable branches

AndrewKushnir commented:
Thanks for the feedback @kara. I've added the logic to fetch upstream and also compare upstream branches, so now the script should compare the latest versions of master and patch branches without additional manual actions. Thank you.

kara reviewed (changes requested):
If I understand your last comment correctly, my local state shouldn't matter now because we should be pulling from upstream branches. But this is what I see under *Only in PATCH*

```
***** Only in PATCH (9.0.x) *****
[v9.0.0-rc.2] bc97c25efe build: add ngJitMode to ng_rollup_bundle terser config (#33865)
[v9.0.0-rc.3] f1bf5b26d1 fix(router): make routerLinkActive work with query params which contain arrays (#22666)
[v9.0.0-rc.5] 1183075d16 Revert "fix(common): update closure locales to include directionality data (#34240)"
[v9.0.0-rc.8] a9f13015b2 build: upgrade to io_bazel_rules_sass 1.24.0 (#34589)
[v9.0.0-rc.8] a812880054 build: package-build & release scripts should handle new pkg_npm naming (#34589)
[v9.0.0-rc.8] d35451327e test: handle bootstrap templated_args in jasmine_node_test defaults.bzl (#34589)
[v9.0.0-rc.8] e175878480 test: keep diff in integration/bazel minimal by not introducing additional_root_paths (#34589)
[v9.0.0-rc.8] f7dc3a7b64 build: renamed "injected_scripts" target to "asset_injected_index_html" (#34589)
[v9.0.0-rc.8] 68e6db5d7a build: add a before-all-other bootstrap script that patches require (#34589)
[v9.0.0-rc.8] d83071c3b1 build: should use $(execpath) in npm_package_bin args (#34589)
[v9.0.0-rc.8] 3e72381bb0 build: derive ts_library dep from jasmine_node_test boostrap label if it ends in `_es5` (#34589)
[v9.0.0-rc.8] c4e9cef314 build: add better comments to defaults.bzl jasmine_node_test macro (#34589)
[v9.0.0-rc.8] 613367508f test: add coverage to  //tools/ts-api-guardian/... on WindowsCI (#34589)
[v9.0.0-rc.10] 20dc436585 fix(language-service): rename getDiagnostics to getSemanticDiagnostics (#34909)
[v9.0.0-rc.11] 8b858fb30a Revert "ci: filter out bad commit from the commit message check to green up master (#35017)" (#35024)
[v9.0.1] ce85cbf2d3 Revert "feat(ngcc): pause async ngcc processing if another process has the lockfile (#35131)"
```

Some of these commits (e.g. a9f13015b2) are also on the master branch, so they shouldn't be here.

AndrewKushnir commented:
@kara, yes local state should not matter. I see the same diff in the output:
```
Comparing branches "9.0.x" and master.

***** Only in MASTER *****
[v9.0.0-rc.4] 3c2438425b feat: add direction property to locale files (#33556)
[v9.0.0-rc.4] 15d3e741e9 feat: update the locale files (#33556)
[v9.0.0] 3c69442dbd feat(compiler-cli): implement NgTscPlugin on top of the NgCompiler API (#34792)
[v9.0.0] daac33cdc8 feat: add basic jest support (#35080)
[v9.0.0] e1160f19be feat: make jasmine patch as zone module (#34676)
[v9.0.0] 332937ef24 feat: make mocha a zone module. (#34719)
[v9.1.0-next.1] 66c06eb1ad fix(language-service): provide completions for the structural directive that only injects the 'ViewContainerRef' (#35466)
[v9.1.0-next.1] 39ef57971c build: refactor zone bazel settings (#34532)
[v9.1.0-next.1] d7d359e3ee feat: support passive event options by defining global variables in zone.js config file (#34503)
[v9.1.0-next.1] e084835fb1 Revert "feat: support passive event options by defining global variables in zone.js config file (#34503)"
[v9.1.0-next.1] af76651ccc refactor: update tscplugin api to match google3 (#35455)
[v9.1.0-next.1] f9d483e76e feat: add a temp solution to support passive event listeners. (#34503)
[v9.1.0-next.1] 4acb676f2e feat: add interface definitions which zone extends EventTarget (#35304)
[v9.1.0-next.1] 500e49f0ac refactor(language-service): normalize hover tests (#35656)
[v9.1.0-next.1] 8f5b7f367c feat(bazel): enable ivy template type-checking in g3 (#35672)
[v9.1.0-next.1] 59c0689ade refactor(ngcc): remove redundant await (#35686)
[v9.0.4] 03d88c7965 feat: define all zone.js configurations to typescript interfaces (#35329)
[v9.0.4] dc40a93317 perf(ngcc): spawn workers lazily (#35719)
[v9.0.4] d7efc45c04 perf(ngcc): only create tasks for non-processed formats (#35719)

***** Only in PATCH (9.0.x) *****
[v9.0.0-rc.2] bc97c25efe build: add ngJitMode to ng_rollup_bundle terser config (#33865)
[v9.0.0-rc.3] f1bf5b26d1 fix(router): make routerLinkActive work with query params which contain arrays (#22666)
[v9.0.0-rc.5] 1183075d16 Revert "fix(common): update closure locales to include directionality data (#34240)"
[v9.0.0-rc.8] a9f13015b2 build: upgrade to io_bazel_rules_sass 1.24.0 (#34589)
[v9.0.0-rc.8] a812880054 build: package-build & release scripts should handle new pkg_npm naming (#34589)
[v9.0.0-rc.8] d35451327e test: handle bootstrap templated_args in jasmine_node_test defaults.bzl (#34589)
[v9.0.0-rc.8] e175878480 test: keep diff in integration/bazel minimal by not introducing additional_root_paths (#34589)
[v9.0.0-rc.8] f7dc3a7b64 build: renamed "injected_scripts" target to "asset_injected_index_html" (#34589)
[v9.0.0-rc.8] 68e6db5d7a build: add a before-all-other bootstrap script that patches require (#34589)
[v9.0.0-rc.8] d83071c3b1 build: should use $(execpath) in npm_package_bin args (#34589)
[v9.0.0-rc.8] 3e72381bb0 build: derive ts_library dep from jasmine_node_test boostrap label if it ends in `_es5` (#34589)
[v9.0.0-rc.8] c4e9cef314 build: add better comments to defaults.bzl jasmine_node_test macro (#34589)
[v9.0.0-rc.8] 613367508f test: add coverage to  //tools/ts-api-guardian/... on WindowsCI (#34589)
[v9.0.0-rc.10] 20dc436585 fix(language-service): rename getDiagnostics to getSemanticDiagnostics (#34909)
[v9.0.0-rc.11] 8b858fb30a Revert "ci: filter out bad commit from the commit message check to green up master (#35017)" (#35024)
[v9.0.1] ce85cbf2d3 Revert "feat(ngcc): pause async ngcc processing if another process has the lockfile (#35131)"
```

These 10 commits that are showing as delta in patch branch are likely artifacts of merging and rolling back PR https://github.com/angular/angular/pull/34589:

```
[v9.0.0-rc.8] a9f13015b2 build: upgrade to io_bazel_rules_sass 1.24.0 (#34589)
[v9.0.0-rc.8] a812880054 build: package-build & release scripts should handle new pkg_npm naming (#34589)
[v9.0.0-rc.8] d35451327e test: handle bootstrap templated_args in jasmine_node_test defaults.bzl (#34589)
[v9.0.0-rc.8] e175878480 test: keep diff in integration/bazel minimal by not introducing additional_root_paths (#34589)
[v9.0.0-rc.8] f7dc3a7b64 build: renamed "injected_scripts" target to "asset_injected_index_html" (#34589)
[v9.0.0-rc.8] 68e6db5d7a build: add a before-all-other bootstrap script that patches require (#34589)
[v9.0.0-rc.8] d83071c3b1 build: should use $(execpath) in npm_package_bin args (#34589)
[v9.0.0-rc.8] 3e72381bb0 build: derive ts_library dep from jasmine_node_test boostrap label if it ends in `_es5` (#34589)
[v9.0.0-rc.8] c4e9cef314 build: add better comments to defaults.bzl jasmine_node_test macro (#34589)
[v9.0.0-rc.8] 613367508f test: add coverage to  //tools/ts-api-guardian/... on WindowsCI (#34589)
```

The compare script is now relies on git to generate deltas with minimum post-processing, i.e. the same 10 commits are present when I execute underlying git command:

```
git log --cherry-pick --oneline --left-only upstream/9.0.x...upstream/master
```

I'll look into this more and keep you updated.

Thank you.

josephperrott commented on scripts/compare-master-to-stable.js:
```diff
@@ -0,0 +1,123 @@
+}
+
+function getAsArray(gitCommand) {
+  return execGitCommand(gitCommand).split('\n');
```
This can for some commands end up with an empty value as the last item in the array.

It might be valuable to `trim` the output of the git command before `split`ing it.

josephperrott commented on scripts/compare-master-to-stable.js:
```diff
@@ -0,0 +1,123 @@
+
+function toMap(input) {
+  let version = 'initial';
+  const data = [];
```
Instead of creating a data array, you could create this as a map from the beginning and then populate the map.

```js
const dataMap = new Map();
input.reverse().forEach((item) => {

...

      dataMap.set(key, [item, version]);
    }
  });
return dataMap;
```

josephperrott commented on scripts/compare-master-to-stable.js:
```diff
@@ -0,0 +1,123 @@
+      .sort(semver.rcompare)[0];
+}
+
+// Main program
```
Can you add comments throughout this main program area explaining what/why things are being done?

josephperrott replied:
We should put the main program inside of a top level function and then call it at the end of the script.

```js
function main() {
  // do things...
}

run();
```

josephperrott commented on scripts/compare-master-to-stable.js:
```diff
@@ -0,0 +1,123 @@
+  return output;
+}
+
+function getAsArray(gitCommand) {
```
Is there another name we can use to express that this is getting the results of a git command as an array?

AndrewKushnir committed:
fixup! build: adding a script to compare commits in master and stable branches

AndrewKushnir commented:
@josephperrott, thanks for the review! I've applied the necessary changes based on your feedback. Could you please take another look when you get a chance? Thank you.

josephperrott commented on scripts/compare-master-to-stable.js:
```diff
@@ -0,0 +1,132 @@
+const ignorePatterns = [
+  'release:',
+  'docs:',
+  'build(docs-infra): upgrade cli command docs',
```
Can we comment/note why we have such a specific pattern we are starting with on this one.

josephperrott commented on scripts/compare-master-to-stable.js:
```diff
@@ -0,0 +1,132 @@
+  return commitsMap;
+}
+
+function diff(mapA, mapB) {
```
You should add documentation here that explains that the values returned here are the items that are in the first provided map, but not in the second.

josephperrott reviewed (approved):
LGTM, in terms of code, Kara will have to give final confirmation that its the correct output we are looking for.

AndrewKushnir committed:
fixup! build: adding a script to compare commits in master and stable branches

AndrewKushnir committed:
fixup! build: adding a script to compare commits in master and stable branches

AndrewKushnir committed:
fixup! build: adding a script to compare commits in master and stable branches

AndrewKushnir commented:
@kara we discussed this Joey: the delta that was showing up in the list was related to the https://github.com/angular/angular/pull/34589 that was rolled back and then the changes were merged again. As a result, for some reasons git was unable to cleanly calculate the delta while using the `--cherry-pick` flag. I've verified that the commits are present in both master and patch.  The current approach is to limit the log starting from the time we cut the 9.0.0 release and see if we have any other discrepancies in the future. As you requested, here is the current diff that I see:

```
***** Only in MASTER *****
[v9.0.0] 3c69442dbd feat(compiler-cli): implement NgTscPlugin on top of the NgCompiler API (#34792)
[v9.0.0] daac33cdc8 feat: add basic jest support (#35080)
[v9.0.0] e1160f19be feat: make jasmine patch as zone module (#34676)
[v9.0.0] 332937ef24 feat: make mocha a zone module. (#34719)
[v9.1.0-next.1] 66c06eb1ad fix(language-service): provide completions for the structural directive that only injects the 'ViewContainerRef' (#35466)
[v9.1.0-next.1] 39ef57971c build: refactor zone bazel settings (#34532)
[v9.1.0-next.1] d7d359e3ee feat: support passive event options by defining global variables in zone.js config file (#34503)
[v9.1.0-next.1] e084835fb1 Revert "feat: support passive event options by defining global variables in zone.js config file (#34503)"
[v9.1.0-next.1] af76651ccc refactor: update tscplugin api to match google3 (#35455)
[v9.1.0-next.1] f9d483e76e feat: add a temp solution to support passive event listeners. (#34503)
[v9.1.0-next.1] 4acb676f2e feat: add interface definitions which zone extends EventTarget (#35304)
[v9.1.0-next.1] 500e49f0ac refactor(language-service): normalize hover tests (#35656)
[v9.1.0-next.1] 8f5b7f367c feat(bazel): enable ivy template type-checking in g3 (#35672)
[v9.1.0-next.1] 59c0689ade refactor(ngcc): remove redundant await (#35686)
[v9.0.4] 03d88c7965 feat: define all zone.js configurations to typescript interfaces (#35329)
[v9.0.4] 4052dd8188 refactor(core): update `noSideEffects()` to return the result of the inner function (#35769)
[v9.0.4] dc6a7918e3 fix(core): add `noSideEffects()` to `make*Decorator()` functions (#35769)
[v9.0.4] ba3612774f fix(core): add `noSideEffects()` to `ɵɵdefineComponent()` (#35769)
[v9.0.4] 9cf85d2177 fix(core): remove side effects from `ɵɵNgOnChangesFeature()` (#35769)
[v9.0.4] c195d22f68 fix(core): remove side effects from `ɵɵgetInheritedFactory()` (#35769)
[v9.0.4] ed44073a58 build(zone.js): sets rollup_bundle to use the --silent flag (#35835)

***** Only in PATCH (9.0.x) *****
[v9.0.1] ce85cbf2d3 Revert "feat(ngcc): pause async ngcc processing if another process has the lockfile (#35131)"
[v9.0.4] 12e52a7696 Revert "build: update to rules_nodejs 1.4.0 (#35813)" (#35836)
```

Which looks correct to me.

Thank you.

IgorMinar commented on scripts/compare-master-to-stable.js:
```diff
@@ -0,0 +1,145 @@
+#!/usr/bin/env node
```
please rename master-to-patch.js (we no longer use the "stable" terminology)

IgorMinar commented on scripts/compare-master-to-stable.js:
```diff
@@ -0,0 +1,145 @@
+// present only in one branch. Ignoring them reduced the "noise" in the final output.
+const ignorePatterns = [
+  'release:',
+  'docs:',
```
why docs? we want to ensure that all docs that should have been cherry-picked were cherry-picked.

AndrewKushnir replied:
I updated that item to only exclude release notes with the `docs:` prefix.

IgorMinar commented on scripts/compare-master-to-stable.js:
```diff
@@ -0,0 +1,145 @@
+  // in `aio/package.json`). Separate commits are generated for master and patch branches and since
+  // it's purely an infrastructure-related change, we ignore these commits while comparing master
+  // and patch diffs to look for delta.
+  'build(docs-infra): upgrade cli command docs sources',
```
aren't we cherry-picking these to the patch branch as well? or do we create separate patch-only PRs for these?

AndrewKushnir replied:
We create separate PRs for master and patch.

AndrewKushnir committed:
fixup! build: adding a script to compare commits in master and stable branches

IgorMinar commented on scripts/compare-master-to-patch.js:
```diff
@@ -0,0 +1,145 @@
+];
+
+// Limit the log history to start from v9.0.0 release date.
+const after = '--after="2020-02-05"';
```
document when this can be removed.

IgorMinar commented on scripts/compare-master-to-patch.js:
```diff
@@ -0,0 +1,145 @@
+  const result = [];
+  mapA.forEach((value, key) => {
+    if (!mapB.has(key)) {
+      result.push(`[${value[1]}] ${value[0]}`);
```
```suggestion
      result.push(`[${value[1]}+] ${value[0]}`);
```

IgorMinar replied:
just so that it's clear that it was merged after the release and no before.

IgorMinar reviewed (approved):
lgtm otherwise. thanks!

AndrewKushnir committed:
fixup! build: adding a script to compare commits in master and stable branches

AndrewKushnir commented:
Thanks for the review @IgorMinar, I've addressed your comments.

IgorMinar commented:
@kara please take a look or remove yourself from the reviewer list if you are happy with our reviews.

kara reviewed (approved):
LGTM
